File: bit_manipulation/single_number.py
from typing import List


class Solution:
    def singleNumber(self, nums: List[int]) -> int:
        hash = {}
        for i in range(len(nums)):
            if nums[i] in hash:
                hash[nums[i]] += 1
            else:
                hash[nums[i]] = 1

        for k, v in hash.items():
            if v == 1:
                return k

    # alternative solution with xor manipulation with O(1) space
        xor = 0
        for i in nums:
            xor ^= i 
        return xor
    

File: bit_manipulation/single_number.go
// Single Number

package main

import "fmt"

func singleNumber(nums []int) int {
	hash := make(map[int]uint8)

	for _, num := range nums {
		hash[num] += 1
	}

	for k, v := range hash {
		if v == 1 {
			return k
		}
	}

	return -1
}

func main() {
	list := []int{4, 1, 2, 1, 2}
	result := singleNumber(list)
	fmt.Println(result)
}
